File: services/backend/app/therapist_search.py
from __future__ import annotations

import math
import time
from typing import Any

import httpx

from .config import settings
from .schemas import TherapistResult

GEOCODE_TIMEOUT = httpx.Timeout(20.0, connect=5.0)
OVERPASS_TIMEOUT = httpx.Timeout(20.0, connect=5.0)
CACHE_TTL_SECONDS = 60 * 15

_CACHE: dict[tuple[str, int], tuple[float, list[TherapistResult]]] = {}
_REQUEST_TIMES: list[float] = []


def clear_cache() -> None:
    _CACHE.clear()
    _REQUEST_TIMES.clear()


def _allow_request() -> bool:
    now = time.time()
    while _REQUEST_TIMES and now - _REQUEST_TIMES[0] > 60:
        _REQUEST_TIMES.pop(0)
    if len(_REQUEST_TIMES) >= 30:
        return False
    _REQUEST_TIMES.append(now)
    return True


def geocode_location(query: str) -> tuple[float, float] | None:
    if not query.strip():
        return None
    try:
        response = httpx.get(
            f"{settings.nominatim_base_url}/search",
            params={"q": query, "format": "json", "limit": 1},
            headers={"User-Agent": settings.therapist_search_user_agent},
            timeout=GEOCODE_TIMEOUT
        )
        response.raise_for_status()
        payload = response.json()
    except httpx.RequestError:
        raise

    if not payload:
        return None
    try:
        lat = float(payload[0]["lat"])
        lon = float(payload[0]["lon"])
    except (KeyError, ValueError, TypeError):
        return None
    return lat, lon


def overpass_search(lat: float, lon: float, radius_m: int) -> list[dict[str, Any]]:
    query = f"""
    [out:json][timeout:25];
    (
      node["healthcare"="psychotherapist"](around:{radius_m},{lat},{lon});
      way["healthcare"="psychotherapist"](around:{radius_m},{lat},{lon});
      relation["healthcare"="psychotherapist"](around:{radius_m},{lat},{lon});
      node["healthcare"="counselling"](around:{radius_m},{lat},{lon});
      way["healthcare"="counselling"](around:{radius_m},{lat},{lon});
      relation["healthcare"="counselling"](around:{radius_m},{lat},{lon});
      node["amenity"="doctors"]["healthcare:speciality"~"psych|psychiatry|psychotherapy",i](around:{radius_m},{lat},{lon});
      way["amenity"="doctors"]["healthcare:speciality"~"psych|psychiatry|psychotherapy",i](around:{radius_m},{lat},{lon});
      relation["amenity"="doctors"]["healthcare:speciality"~"psych|psychiatry|psychotherapy",i](around:{radius_m},{lat},{lon});
    );
    out center tags;
    """
    try:
        response = httpx.post(
            f"{settings.overpass_base_url}/api/interpreter",
            content=query,
            headers={"User-Agent": settings.therapist_search_user_agent},
            timeout=OVERPASS_TIMEOUT
        )
        response.raise_for_status()
        payload = response.json()
    except httpx.RequestError:
        raise

    return payload.get("elements", [])


def _format_address(tags: dict[str, Any]) -> str:
    full = tags.get("addr:full")
    if full:
        return full
    parts: list[str] = []
    street = tags.get("addr:street")
    number = tags.get("addr:housenumber")
    if street:
        parts.append(f"{street} {number}".strip() if number else street)
    city = tags.get("addr:city")
    postcode = tags.get("addr:postcode")
    locality = " ".join([part for part in [postcode, city] if part])
    if locality:
        parts.append(locality)
    return ", ".join(parts) if parts else "Address unavailable"


def _haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    radius = 6371.0
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)
    a = math.sin(delta_phi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2) ** 2
    return radius * (2 * math.atan2(math.sqrt(a), math.sqrt(1 - a)))


def normalize_providers(
    elements: list[dict[str, Any]],
    origin_lat: float,
    origin_lon: float,
    limit: int
) -> list[TherapistResult]:
    providers: list[TherapistResult] = []
    for element in elements:
        lat = element.get("lat") or element.get("center", {}).get("lat")
        lon = element.get("lon") or element.get("center", {}).get("lon")
        if lat is None or lon is None:
            continue
        tags = element.get("tags") or {}
        name = tags.get("name") or tags.get("brand") or "Therapist"
        address = _format_address(tags)
        phone = tags.get("phone") or tags.get("contact:phone") or ""
        website = tags.get("website") or tags.get("contact:website")
        osm_url = f"https://www.openstreetmap.org/{element.get('type')}/{element.get('id')}"
        url = website or osm_url
        distance_km = _haversine_km(origin_lat, origin_lon, float(lat), float(lon))
        providers.append(
            TherapistResult(
                name=name,
                address=address,
                url=url,
                phone=phone,
                distance_km=round(distance_km, 2)
            )
        )
    providers.sort(key=lambda provider: provider.distance_km)
    return providers[:limit]


def _demo_providers(query: str) -> list[TherapistResult]:
    return [
        TherapistResult(
            name=f"{query.title()} Counseling",
            address=f"{query.title()} Center",
            url="https://www.openstreetmap.org/",
            phone="",
            distance_km=0.5
        )
    ]


def search_therapists(query: str, radius_km: int | None = None) -> list[TherapistResult]:
    normalized_query = query.strip().lower()
    if not settings.therapist_search_enabled:
        return []
    radius = radius_km or settings.therapist_search_radius_km_default
    radius_m = int(radius * 1000)
    cache_key = (normalized_query, radius)
    now = time.time()
    cached = _CACHE.get(cache_key)
    if cached and now - cached[0] < CACHE_TTL_SECONDS:
        return cached[1]

    if not _allow_request():
        return []

    try:
        coords = geocode_location(query)
        if not coords:
            return _demo_providers(query) if settings.demo_mode else []
        lat, lon = coords
        elements = overpass_search(lat, lon, radius_m)
        providers = normalize_providers(elements, lat, lon, limit=settings.therapist_search_limit)
        if not providers and settings.demo_mode:
            providers = _demo_providers(query)
        _CACHE[cache_key] = (now, providers)
        return providers
    except httpx.RequestError:
        return []

File: apps/frontend/app/api/_lib/therapist-search.ts
import Ajv from "ajv";
import type { TherapistResult } from "./safety";

const ajv = new Ajv({ allErrors: true });

const nominatimSchema = {
  type: "array",
  items: {
    type: "object",
    properties: {
      lat: { type: "string" },
      lon: { type: "string" }
    },
    required: ["lat", "lon"],
    additionalProperties: true
  }
} as const;

const overpassSchema = {
  type: "object",
  properties: {
    elements: {
      type: "array",
      items: {
        type: "object",
        properties: {
          type: { type: "string" },
          id: { type: "number" },
          lat: { type: "number" },
          lon: { type: "number" },
          center: {
            type: "object",
            properties: {
              lat: { type: "number" },
              lon: { type: "number" }
            },
            required: ["lat", "lon"],
            additionalProperties: true
          },
          tags: { type: "object" }
        },
        required: ["type", "id"],
        additionalProperties: true
      }
    }
  },
  required: ["elements"],
  additionalProperties: true
} as const;

const validateNominatim = ajv.compile(nominatimSchema);
const validateOverpass = ajv.compile(overpassSchema);

const CACHE_TTL_MS = 10 * 60 * 1000;
const REQUEST_LIMIT = 30;

const cache = new Map<string, { expiresAt: number; results: TherapistResult[] }>();
const requestTimes: number[] = [];

const nominatimBase = (process.env.NOMINATIM_BASE_URL || "https://nominatim.openstreetmap.org").replace(/\/$/, "");
const overpassBase = (process.env.OVERPASS_BASE_URL || "https://overpass-api.de").replace(/\/$/, "");
const parsedRadius = Number(process.env.THERAPIST_SEARCH_RADIUS_KM_DEFAULT || 10);
const parsedLimit = Number(process.env.THERAPIST_SEARCH_LIMIT || 10);
const defaultRadiusKm = Number.isFinite(parsedRadius) ? parsedRadius : 10;
const limit = Number.isFinite(parsedLimit) ? parsedLimit : 10;
const userAgent = process.env.THERAPIST_SEARCH_USER_AGENT || "mh-skills-coach/0.1 (vercel)";

const allowRequest = () => {
  const now = Date.now();
  while (requestTimes.length > 0 && now - requestTimes[0] > 60_000) {
    requestTimes.shift();
  }
  if (requestTimes.length >= REQUEST_LIMIT) {
    return false;
  }
  requestTimes.push(now);
  return true;
};

const fetchWithTimeout = async (url: string, init: RequestInit, timeoutMs: number) => {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...init, signal: controller.signal });
  } finally {
    clearTimeout(timeout);
  }
};

const nominatimEndpoint = () => {
  if (nominatimBase.endsWith("/search")) {
    return nominatimBase;
  }
  return `${nominatimBase}/search`;
};

const overpassEndpoint = () => {
  if (overpassBase.endsWith("/api/interpreter")) {
    return overpassBase;
  }
  return `${overpassBase}/api/interpreter`;
};

const geocodeLocation = async (query: string): Promise<[number, number] | null> => {
  if (!query.trim()) {
    return null;
  }
  const url = new URL(nominatimEndpoint());
  url.searchParams.set("q", query);
  url.searchParams.set("format", "json");
  url.searchParams.set("limit", "1");

  const res = await fetchWithTimeout(
    url.toString(),
    {
      headers: {
        "User-Agent": userAgent,
        Accept: "application/json"
      }
    },
    10_000
  );

  if (!res.ok) {
    return null;
  }
  const payload: unknown = await res.json();
  if (!validateNominatim(payload)) {
    return null;
  }
  const results = payload as Array<{ lat: string; lon: string }>;
  if (results.length === 0) {
    return null;
  }
  const lat = Number(results[0]?.lat);
  const lon = Number(results[0]?.lon);
  if (!Number.isFinite(lat) || !Number.isFinite(lon)) {
    return null;
  }
  return [lat, lon];
};

const overpassSearch = async (lat: number, lon: number, radiusM: number) => {
  const query = `
  [out:json][timeout:25];
  (
    node["healthcare"="psychotherapist"](around:${radiusM},${lat},${lon});
    way["healthcare"="psychotherapist"](around:${radiusM},${lat},${lon});
    relation["healthcare"="psychotherapist"](around:${radiusM},${lat},${lon});
    node["healthcare"="psychologist"](around:${radiusM},${lat},${lon});
    way["healthcare"="psychologist"](around:${radiusM},${lat},${lon});
    relation["healthcare"="psychologist"](around:${radiusM},${lat},${lon});
    node["healthcare"="psychiatrist"](around:${radiusM},${lat},${lon});
    way["healthcare"="psychiatrist"](around:${radiusM},${lat},${lon});
    relation["healthcare"="psychiatrist"](around:${radiusM},${lat},${lon});
    node["healthcare"="counselling"](around:${radiusM},${lat},${lon});
    way["healthcare"="counselling"](around:${radiusM},${lat},${lon});
    relation["healthcare"="counselling"](around:${radiusM},${lat},${lon});
    node["amenity"="clinic"]["healthcare:speciality"~"psych|psychiatry|psychotherapy",i](around:${radiusM},${lat},${lon});
    way["amenity"="clinic"]["healthcare:speciality"~"psych|psychiatry|psychotherapy",i](around:${radiusM},${lat},${lon});
    relation["amenity"="clinic"]["healthcare:speciality"~"psych|psychiatry|psychotherapy",i](around:${radiusM},${lat},${lon});
  );
  out center tags;
  `;

  const res = await fetchWithTimeout(
    overpassEndpoint(),
    {
      method: "POST",
      headers: {
        "User-Agent": userAgent,
        Accept: "application/json"
      },
      body: query
    },
    25_000
  );

  if (res.status === 429 || res.status >= 500) {
    return [];
  }
  if (!res.ok) {
    return [];
  }
  const payload: unknown = await res.json();
  if (!validateOverpass(payload)) {
    return [];
  }
  const data = payload as { elements: Array<Record<string, unknown>> };
  return data.elements;
};

const formatAddress = (tags: Record<string, unknown>) => {
  const full = tags["addr:full"];
  if (typeof full === "string" && full.trim()) {
    return full;
  }
  const parts: string[] = [];
  const street = tags["addr:street"];
  const number = tags["addr:housenumber"];
  if (typeof street === "string" && street.trim()) {
    const line = typeof number === "string" && number.trim() ? `${street} ${number}` : street;
    parts.push(line);
  }
  const city = typeof tags["addr:city"] === "string" ? tags["addr:city"] : "";
  const postcode = typeof tags["addr:postcode"] === "string" ? tags["addr:postcode"] : "";
  const country =
    (typeof tags["addr:country"] === "string" && tags["addr:country"]) ||
    (typeof tags["addr:country_code"] === "string" && tags["addr:country_code"]) ||
    (typeof tags["country"] === "string" && tags["country"]) ||
    "";
  const locality = [postcode, city].filter(Boolean).join(" ");
  if (locality) {
    parts.push(locality);
  }
  if (!parts.length && (city || country)) {
    return [city, country].filter(Boolean).join(", ");
  }
  if (!street && locality && country) {
    parts.push(country);
  }
  return parts.length ? parts.join(", ") : "Address unavailable";
};

const haversineKm = (lat1: number, lon1: number, lat2: number, lon2: number) => {
  const radius = 6371;
  const phi1 = (lat1 * Math.PI) / 180;
  const phi2 = (lat2 * Math.PI) / 180;
  const deltaPhi = ((lat2 - lat1) * Math.PI) / 180;
  const deltaLambda = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(deltaPhi / 2) ** 2 +
    Math.cos(phi1) * Math.cos(phi2) * Math.sin(deltaLambda / 2) ** 2;
  return radius * (2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a)));
};

const normalizeProviders = (
  elements: Array<Record<string, unknown>>,
  originLat: number,
  originLon: number
): TherapistResult[] => {
  const providers: TherapistResult[] = [];
  for (const element of elements) {
    const lat =
      typeof element.lat === "number"
        ? element.lat
        : typeof element.center === "object" && element.center
          ? (element.center as { lat?: number }).lat
          : undefined;
    const lon =
      typeof element.lon === "number"
        ? element.lon
        : typeof element.center === "object" && element.center
          ? (element.center as { lon?: number }).lon
          : undefined;
    if (!Number.isFinite(lat) || !Number.isFinite(lon)) {
      continue;
    }
    const tags =
      typeof element.tags === "object" && element.tags ? (element.tags as Record<string, unknown>) : {};
    const name =
      (typeof tags.name === "string" && tags.name) ||
      (typeof tags.brand === "string" && tags.brand) ||
      "Therapist";
    const address = formatAddress(tags);
    const phone =
      (typeof tags.phone === "string" && tags.phone) ||
      (typeof tags["contact:phone"] === "string" && tags["contact:phone"]) ||
      "Phone unavailable";
    const website =
      (typeof tags.website === "string" && tags.website) ||
      (typeof tags["contact:website"] === "string" && tags["contact:website"]) ||
      "";
    const type = typeof element.type === "string" ? element.type : "node";
    const id = typeof element.id === "number" ? element.id : 0;
    const osmUrl = `https://www.openstreetmap.org/${type}/${id}`;
    const url = website || osmUrl;
    const distanceKm = haversineKm(originLat, originLon, Number(lat), Number(lon));
    providers.push({
      name,
      address,
      url,
      phone,
      distance_km: Math.round(distanceKm * 100) / 100
    });
  }
  providers.sort((a, b) => a.distance_km - b.distance_km);
  return providers.slice(0, limit);
};

export const searchTherapists = async (query: string, radiusKm?: number) => {
  const trimmedQuery = query.trim();
  if (!trimmedQuery) {
    return [];
  }
  const radius = Number.isFinite(radiusKm) && radiusKm ? radiusKm : defaultRadiusKm;
  if (!allowRequest()) {
    return [];
  }
  try {
    const coords = await geocodeLocation(trimmedQuery);
    if (!coords) {
      return [];
    }
    const [lat, lon] = coords;
    const cacheKey = `${lat.toFixed(3)}:${lon.toFixed(3)}:${radius}`;
    const now = Date.now();
    const cached = cache.get(cacheKey);
    if (cached && cached.expiresAt > now) {
      return cached.results;
    }
    const elements = await overpassSearch(lat, lon, Math.round(radius * 1000));
    const providers = normalizeProviders(elements, lat, lon);
    cache.set(cacheKey, { expiresAt: now + CACHE_TTL_MS, results: providers });
    return providers;
  } catch {
    return [];
  }
};
